// @h4ad/serverless-adapter/lib/resolvers/aws-context/aws-context.resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AwsContextResolver = void 0;
const { __立ResolverProps } = require('../../contracts');
const { __立Resolver } = require('../../contracts');
function __assignType(fn, args) {
    fn.__type = args;
    return fn;
}
//#endregion
/**
 * The class that implements the resolver by using the AWS Context object.
 *
 * @remarks To use this resolver, you MUST leave `{@link https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html | callbackWaitsForEmptyEventLoop}` as true, otherwise, AWS will not wait for this resolver to resolve.
 *
 * @deprecated From the AWS Documentation, describing the functions used in this resolver: Functions for compatibility with earlier Node.js Runtime v0.10.42. No longer documented, so they are deprecated, but they still work as of the 12.x runtime, so they are not removed from the types.
 *
 * @breadcrumb Resolvers / AwsContextResolver
 * @public
 */
class AwsContextResolver {
    /**
     * {@inheritDoc}
     */
    createResolver({ context, event, log, respondWithErrors, adapter, }) {
        if (!context) {
            throw new Error('Could not figure out how to create the resolver because the "context" argument was not sent.');
        }
        if (!context.succeed) {
            throw new Error('Could not figure out how to create the resolver because the "context" argument didn\'t have the "succeed" function.');
        }
        if (!context.fail) {
            throw new Error('Could not figure out how to create the resolver because the "context" argument didn\'t have the "fail" function.');
        }
        const delegatedResolver = {
            succeed: __assignType(response => context.succeed(response), ['response', '', 'P"2!"/"']),
            fail: __assignType(error => context.fail(error), ['error', '', 'P"2!"/"']),
        };
        return {
            run: __assignType(task => {
                task()
                    .then(__assignType(response => delegatedResolver.succeed(response), ['response', '', 'P"2!"/"']))
                    .catch(__assignType(error => {
                    log.error('SERVERLESS_ADAPTER:RESPOND_TO_EVENT_SOURCE_WITH_ERROR', error);
                    adapter.onErrorWhileForwarding({
                        delegatedResolver,
                        error,
                        log,
                        event,
                        respondWithErrors,
                    });
                }, ['error', '', 'P"2!"/"']));
            }, ['task', '', 'P"2!"/"']),
        };
    }
}
exports.AwsContextResolver = AwsContextResolver;
AwsContextResolver.__type = ['TEvent', 'TCallback', 'TResponse', () => __立ResolverProps, 'param0', () => __立Resolver, 'createResolver', 'b!b"b#P""""o$%2%"$o&#0\'5e!!"e!"e!#$6&'];
