// @h4ad/serverless-adapter/lib/handlers/huawei/http-huawei.handler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpHuaweiHandler = exports.__ΩHttpHuaweiHandlerOptions = exports.DEFAULT_HUAWEI_LISTEN_PORT = void 0;
const { __ΩFrameworkContract } = require('../../contracts');
const { __ΩAdapterContract } = require('../../contracts');
const { __ΩResolverContract } = require('../../contracts');
const { __ΩBinarySettings } = require('../../@types');
const { __ΩILogger } = require('../../core');
const { __ΩServerlessHandler } = require('../../contracts');
function __assignType(fn, args) {
    fn.__type = args;
    return fn;
}
//#region Imports
const http = __importStar(require("http"));
const core_1 = require("../../core");
//#endregion
/**
 * The default port that huawei will proxy the request to your framework
 *
 * {@link https://support.huaweicloud.com/intl/en-us/ae-ad-1-usermanual-functiongraph/functiongraph_01_1442.html#functiongraph_01_1442__li194597302096 | Reference}
 *
 * @breadcrumb Handlers / HttpHuaweiHandler
 * @public
 */
exports.DEFAULT_HUAWEI_LISTEN_PORT = 8000;
const __ΩHttpHuaweiHandlerOptions = ['port', 'requestListener', '', 'httpServerFactory', 'P\'4!8P!2"!/#4$8M'];
exports.__ΩHttpHuaweiHandlerOptions = __ΩHttpHuaweiHandlerOptions;
/**
 * The class that implements a huawei serverless handler with http function that exposes a http server in specific port.
 *
 * In this Handler, you don't need to specific resolver and adapter, so you can use DummyAdapter and DummyResolver instead.
 *
 * @see https://support.huaweicloud.com/intl/en-us/ae-ad-1-usermanual-functiongraph/functiongraph_01_1442.html#functiongraph_01_1442__li194597302096
 *
 * @breadcrumb Handlers / HttpHuaweiHandler
 * @public
 */
class HttpHuaweiHandler {
    //#region Constructor
    /**
     * Construtor padrão
     */
    constructor(options) {
        this.options = options;
    }
    //#endregion
    //#region Public Methods
    /**
     * {@inheritDoc}
     */
    getHandler(app, framework, _, __, binarySettings, respondWithErrors, log) {
        var _a, _b;
        const requestListener = __assignType((req, res) => {
            framework.sendRequest(app, req, res);
        }, ['req', 'res', '', 'P"2!"2""/#']);
        const server = (0, core_1.getDefaultIfUndefined)((_a = this.options) === null || _a === void 0 ? void 0 : _a.httpServerFactory, this.createHttpServer.bind(this))(requestListener);
        const port = (0, core_1.getDefaultIfUndefined)((_b = this.options) === null || _b === void 0 ? void 0 : _b.port, exports.DEFAULT_HUAWEI_LISTEN_PORT);
        server.listen(port, () => {
            log.debug('SERVERLESS_ADAPTER:PROXY: Server started.');
            log.debug(`SERVERLESS_ADAPTER:PROXY: Listening the Huawei Requests in port ${port}`);
        });
        return () => {
            log.debug('SERVERLESS_ADAPTER:PROXY: Disposing the server');
            return new Promise(__assignType((resolve, reject) => {
                server.close(__assignType(err => (err ? reject(err) : resolve()), ['err', '', 'P"2!"/"']));
            }, ['resolve', 'reject', '', 'P"2!"2""/#']));
        };
    }
    //#endregion
    //#region Protected Methods
    /**
     * The method that creates the instance of http server
     *
     * @param requestListener - O método que lidará com as requisições recebidas
     */
    createHttpServer(requestListener) {
        return http.createServer(requestListener);
    }
}
exports.HttpHuaweiHandler = HttpHuaweiHandler;
HttpHuaweiHandler.__type = ['TApp', () => __ΩHttpHuaweiHandlerOptions, 'options', 'constructor', 'app', () => __ΩFrameworkContract, 'framework', () => __ΩAdapterContract, '_', () => __ΩResolverContract, '__', () => __ΩBinarySettings, 'binarySettings', 'respondWithErrors', () => __ΩILogger, 'log', () => __ΩServerlessHandler, 'getHandler', 'requestListener', 'createHttpServer', 'b!Pn"2#8<9"0$Pe"!2%e"!o&"2\'$$$o($F2)$$$$$o*&2+n,2-)2.n/20$`o1"02P!23!04<5e!!$$$$$`6\''];
