// @h4ad/serverless-adapter/lib/adapters/aws/sns.adapter.js
"use strict";
//#region Imports
Object.defineProperty(exports, "__esModule", { value: true });
exports.SNSAdapter = exports.__ΩSNSAdapterOptions = void 0;
const core_1 = require("../../core");
const index_1 = require("./base/index");
const __ΩSNSAdapterOptions = ['snsForwardPath', 'snsForwardMethod', 'P&4!8&4"8M'];
exports.__ΩSNSAdapterOptions = __ΩSNSAdapterOptions;
/**
 * The adapter to handle requests from AWS SNS.
 *
 * The option of `responseWithErrors` is ignored by this adapter and we always call `resolver.fail` with the error.
 *
 * {@link https://docs.aws.amazon.com/pt_br/lambda/latest/dg/with-sns.html | Event Reference}
 *
 * @example
 * ```typescript
 * const snsForwardPath = '/your/route/sns'; // default /sns
 * const snsForwardMethod = 'POST'; // default POST
 * const adapter = new SNSAdapter({ snsForwardPath, snsForwardMethod });
 * ```
 *
 * @breadcrumb Adapters / AWS / SNSAdapter
 * @public
 */
class SNSAdapter extends index_1.AwsSimpleAdapter {
    //#region Constructor
    /**
     * Default constructor
     *
     * @param options - The options to customize the {@link SNSAdapter}
     */
    constructor(options) {
        super({
            forwardPath: (0, core_1.getDefaultIfUndefined)(options === null || options === void 0 ? void 0 : options.snsForwardPath, '/sns'),
            forwardMethod: (0, core_1.getDefaultIfUndefined)(options === null || options === void 0 ? void 0 : options.snsForwardMethod, 'POST'),
            batch: false,
            host: 'sns.amazonaws.com',
        });
    }
    //#endregion
    //#region Public Methods
    /**
     * {@inheritDoc}
     */
    getAdapterName() {
        return SNSAdapter.name;
    }
    /**
     * {@inheritDoc}
     */
    canHandle(event) {
        var _a;
        const snsEvent = event;
        if (!Array.isArray(snsEvent === null || snsEvent === void 0 ? void 0 : snsEvent.Records))
            return false;
        const eventSource = (_a = snsEvent.Records[0]) === null || _a === void 0 ? void 0 : _a.EventSource;
        return eventSource === 'aws:sns';
    }
}
exports.SNSAdapter = SNSAdapter;
SNSAdapter.__type = [() => index_1.AwsSimpleAdapter, () => __ΩSNSAdapterOptions, 'options', 'constructor', 'getAdapterName', 'event', 'canHandle', 'P"7!Pn"2#8"0$P&0%P#2&!0\'5"6"'];
