// @h4ad/serverless-adapter/lib/adapters/aws/sqs.adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQSAdapter = exports.__ΩSQSAdapterOptions = void 0;
const { __ΩAWSSimpleAdapterOptions } = require('./base/index');
const __ΩPick = ['T', 'K', 'l+e#!e"!fRb!b"Pde""N#!'];
const core_1 = require("../../core");
const index_1 = require("./base/index");
const __ΩSQSAdapterOptions = [() => __ΩPick, () => __ΩAWSSimpleAdapterOptions, "batch", 'sqsForwardPath', 'sqsForwardMethod', 'Pn".#o!#&4$8&4%8M'];
exports.__ΩSQSAdapterOptions = __ΩSQSAdapterOptions;
/**
 * The adapter to handle requests from AWS SQS.
 *
 * The option of `responseWithErrors` is ignored by this adapter and we always call `resolver.fail` with the error.
 *
 * {@link https://docs.aws.amazon.com/pt_br/lambda/latest/dg/with-sqs.html | Event Reference}
 *
 * @example
 * ```typescript
 * const sqsForwardPath = '/your/route/sqs'; // default /sqs
 * const sqsForwardMethod = 'POST'; // default POST
 * const adapter = new SQSAdapter({ sqsForwardPath, sqsForwardMethod });
 * ```
 *
 * @breadcrumb Adapters / AWS / SQSAdapter
 * @public
 */
class SQSAdapter extends index_1.AwsSimpleAdapter {
    //#region Constructor
    /**
     * Default constructor
     *
     * @param options - The options to customize the {@link SNSAdapter}
     */
    constructor(options) {
        super({
            forwardPath: (0, core_1.getDefaultIfUndefined)(options === null || options === void 0 ? void 0 : options.sqsForwardPath, '/sqs'),
            forwardMethod: (0, core_1.getDefaultIfUndefined)(options === null || options === void 0 ? void 0 : options.sqsForwardMethod, 'POST'),
            batch: options === null || options === void 0 ? void 0 : options.batch,
            host: 'sqs.amazonaws.com',
        });
    }
    //#endregion
    //#region Public Methods
    /**
     * {@inheritDoc}
     */
    getAdapterName() {
        return SQSAdapter.name;
    }
    /**
     * {@inheritDoc}
     */
    canHandle(event) {
        var _a;
        const sqsEvent = event;
        if (!Array.isArray(sqsEvent === null || sqsEvent === void 0 ? void 0 : sqsEvent.Records))
            return false;
        const eventSource = (_a = sqsEvent.Records[0]) === null || _a === void 0 ? void 0 : _a.eventSource;
        return eventSource === 'aws:sqs';
    }
}
exports.SQSAdapter = SQSAdapter;
SQSAdapter.__type = [() => index_1.AwsSimpleAdapter, () => __ΩSQSAdapterOptions, 'options', 'constructor', 'getAdapterName', 'event', 'canHandle', 'P"7!Pn"2#8"0$P&0%P#2&!0\'5"6"'];
