// @h4ad/serverless-adapter/lib/network/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerlessRequest = exports.__ΩServerlessRequestProps = void 0;
const { __ΩSingleValueHeaders } = require('../@types');
// ATTRIBUTION: https://github.com/dougmoscrop/serverless-http
const http_1 = require("http");
const core_1 = require("../core");
const HTTPS_PORT = 443;
const __ΩServerlessRequestProps = ['method', 'url', () => __ΩSingleValueHeaders, 'headers', 'body', 'remoteAddress', 'P&4!&4"n#4$P!WJ4%8&4&8M'];
exports.__ΩServerlessRequestProps = __ΩServerlessRequestProps;
/**
 * The class that represents an {@link http#IncomingMessage} created by the library to represent an actual request to the framework.
 *
 * @breadcrumb Network / ServerlessRequest
 * @public
 */
class ServerlessRequest extends http_1.IncomingMessage {
    constructor({ method, url, headers, body, remoteAddress, }) {
        super({
            encrypted: true,
            readable: false,
            remoteAddress,
            address: () => ({ port: HTTPS_PORT }),
            end: core_1.NO_OP,
            destroy: core_1.NO_OP,
        });
        this.statusCode = 200;
        this.statusMessage = 'OK';
        this.complete = true;
        this.httpVersion = '1.1';
        this.httpVersionMajor = 1;
        this.httpVersionMinor = 1;
        this.method = method;
        this.headers = headers;
        this.body = body;
        this.url = url;
        this.ip = remoteAddress;
        this._read = () => {
            this.push(body);
            this.push(null);
        };
    }
}
exports.ServerlessRequest = ServerlessRequest;
ServerlessRequest.__type = [() => http_1.IncomingMessage, () => __ΩServerlessRequestProps, 'param0', 'constructor', 'ip', 'body', 'P7!Pn"2#"0$&3%8P!WJ3&85'];
